### PR TITLE
Homework4 - Emag tests

### PR DESCRIPTION
**Pull request by MariBuleandra**
Deleted the existing content for EmagOnClassTests.java and added the following:
- [X] register tests
- [X] add a product to the shopping cart
- [X] add multiple products to favourites

### CODE REVIEW

**Comment by MariBuleandra** [comment]
Hello, the following updates were made:
- [X] pulled changes from master to homework4 branch;
- [X] added new methods to helpers/WebDriverHelpers;
- [X] added new constants to pageObjects/emag/CommonPage, pageObjects/emag//FavoritesPage, pageObjects/emag/MainPage;
- [X] added new classes in pageObjects/emag: LoginPage, PDPPage, RegisterPage;
- [X] added new classes in steps.emag: AddToCartSteps, AddToFavouritesSteps, SignUpSteps;
- [X] added new classes in tests/emag.homepageTests: AddToCartTests, AddToFavouritesTest, SignUpTests (with the corresponding code for tests);

Please note that the useless EmagOnClassTests class was not deleted for didactic purpose.

File: src/test/java/helpers/WebDriverHelpers.java [hunk 1]
```diff
@@ -92,4 +94,24 @@ public static void forceClickWebElement (By byObject) {
         js.executeScript("arguments[0].click()", chromeDriver.findElement(byObject));
     }
 
+    public static void hoverOverWebElement (By byObject) {
+        Actions actions = new Actions(chromeDriver);
+        actions.moveToElement(chromeDriver.findElement(byObject)).perform();
```
Comment (BojteTamas) [inline src/test/java/helpers/WebDriverHelpers.java:99]: Please add `build`

File: src/test/java/helpers/WebDriverHelpers.java [hunk 2]
```diff
@@ -92,4 +94,24 @@ public static void forceClickWebElement (By byObject) {
...
+    public static List<WebElement> getListOfElements(By byObject){
+        waitUntilWebElementIsPresent(byObject);
+        return chromeDriver.findElements(byObject);
+    }
+
+    public static void selectFromListByIndex(int index, By byObject) {
+        List<WebElement> list = getListOfElements(byObject);
+        list.get(index).click();
+    }
+
```
Comment (BojteTamas) [inline src/test/java/helpers/WebDriverHelpers.java:116]: To add `selectFromListByText`

File: src/test/java/steps/emag/CommonSteps.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import pageObjects.emag.CommonPage;
+
+public class CommonSteps extends WebDriverHelpers {
+
+    public static void hoverOverMyAccountButton () {
+        hoverOverWebElement(CommonPage.MY_ACCOUNT_BUTTON);
+    }
+
+    public static void checkMyAccountMessage () {
+        checkWebElementString((CommonPage.MY_ACCOUNT_MESSAGE), "Intra in contul tau eMAG si ai control complet asupra ofertelor!");
```
Comment (BojteTamas) [inline src/test/java/steps/emag/CommonSteps.java:13]: Please remove the parentheses. `(CommonPage.MY_ACCOUNT_MESSAGE)`

File: src/test/java/steps/emag/CommonSteps.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    }
+
+    public static void clickSignUpButton () {
+        forceClickWebElement(CommonPage.MY_ACCOUNT_SIGN_UP_BUTTON);
+    }
+
+    public static void checkCartCount () {
```
Comment (BojteTamas) [inline src/test/java/steps/emag/CommonSteps.java:20]: Please add the count as parameter

File: src/test/java/steps/emag/CommonSteps.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        checkWebElementString(CommonPage.COMMON_SHOPPING_CART_PRODUCTS_COUNT, "1");
+    }
+
+    public static void clickFavouritesButton () {
```
Comment (BojteTamas) [inline src/test/java/steps/emag/CommonSteps.java:24]: Please use US favorite

File: src/test/java/steps/emag/FavouriteSteps.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    static String urlPart = "favorites";
+
+    public static void checkFavouritesUrl () {
+        url = chromeDriver.getCurrentUrl();
+        Assert.assertTrue(url.contains(urlPart), "The url doesn't contain" + urlPart);
+
+    }
+
+    public static void checkFavouritesPageTitle () {
+        checkWebElementString(FavoritesPage.PRODUCTS_COUNT, "3 produse");
```
Comment (BojteTamas) [inline src/test/java/steps/emag/FavouriteSteps.java:18]: Please add the title as parameter or the number only

File: src/test/java/steps/emag/LoginSteps.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        String inputEmailValue = getAttributeFromWebElement(LoginPage.LOGIN_EMAIL_INPUT, "value");
+        Assert.assertEquals(email, inputEmailValue);
+    }
+
+    public static void checkLoginErrorMessageNotDisplayed () {
+        checkIfTheElementIsNotPresent(LoginPage.LOGIN_ERROR_MESSAGE);
+    }
+
+    public static void clickContinuaButton () {
+        clickWebElement(LoginPage.LOGIN_SUBMIT_BUTTON);
```
Comment (BojteTamas) [inline src/test/java/steps/emag/LoginSteps.java:30]: Please take a look on the selector name

File: src/test/java/steps/emag/MainSteps.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
+package steps.emag;
+
+import helpers.WebDriverHelpers;
+import pageObjects.emag.CommonPage;
+import pageObjects.emag.MainPage;
+
+public class MainSteps extends WebDriverHelpers {
+
+    public static void clickTheSecondProductFromFirstCarousel () {
```
Comment (BojteTamas) [inline src/test/java/steps/emag/MainSteps.java:9]: please add the index as parameter, to reuse the method.

File: src/test/java/steps/emag/MainSteps.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    public static void clickTheFavouriteButtonFirstProduct () {
+        forceClickWebElement(MainPage.FIRST_CONTAINER_FAVOURITE_BUTTON);
+    }
+
+    public static void clickTheFavouriteButtonSecondProduct () {
+        forceClickWebElement(MainPage.SECOND_CONTAINER_FAVOURITE_BUTTON);
+    }
+
+    public static void clickTheFavouriteButtonThirdProduct () {
+        forceClickWebElement(MainPage.THIRD_CONTAINER_FAVOURITE_BUTTON);
```
Comment (BojteTamas) [inline src/test/java/steps/emag/MainSteps.java:22]: Please remove this methods and use one more general one.

File: src/test/java/steps/emag/MainSteps.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    public static void checkNumberOfFavourites () {
+        chromeDriver.navigate().refresh();
```
Comment (BojteTamas) [inline src/test/java/steps/emag/MainSteps.java:26]: Please remove the refresh

File: src/test/java/steps/emag/MainSteps.java [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+        waitUntilWebElementIsClickable(CommonPage.COMMON_FAVOURITE_PRODUCTS_COUNT);
+        checkWebElementString(CommonPage.COMMON_FAVOURITE_PRODUCTS_COUNT, "3");
```
Comment (BojteTamas) [inline src/test/java/steps/emag/MainSteps.java:28]: Please add the "3" as int parameter

File: src/test/java/steps/emag/PDPSteps.java [hunk 12]
```diff
@@ -1,18 +1,15 @@
 package steps.emag;
 
 import helpers.WebDriverHelpers;
-import org.testng.Assert;
-import pageObjects.emag.CommonPage;
-import pageObjects.emag.MainPage;
 import pageObjects.emag.PDPPage;
 
-public class AddToCartSteps extends WebDriverHelpers {
+public class PDPSteps extends WebDriverHelpers {
```
Comment (BojteTamas) [inline src/test/java/steps/emag/PDPSteps.java:6]: Please use camelCase on the class name

File: src/test/java/steps/emag/PDPSteps.java [hunk 13]
```diff
@@ -1,18 +1,15 @@
...
 
-    public static void clickTheSecondElement () {
-        clickWebElement(MainPage.SECOND_PRODUCT_FROM_FIRST_CONTAINER);
+    public static void checkPDPProductCode () {
```
Comment (BojteTamas) [inline src/test/java/steps/emag/PDPSteps.java:8]: Please add what are you checking exactly

File: src/test/java/steps/emag/RegisterSteps.java [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+        name = RandomStringUtils.randomAlphabetic(7).toUpperCase();
+        fillWebElement(RegisterPage.REGISTER_NAME_INPUT, name);
+    }
+
+    public static void checkRegisterNameFieldInput () {
+        String registerNameInputValue = getAttributeFromWebElement(RegisterPage.REGISTER_NAME_INPUT, "value");
+        Assert.assertEquals(name, registerNameInputValue);
+    }
+
+    public static void checkRegisterNameErrorMessage () {
```
Comment (BojteTamas) [inline src/test/java/steps/emag/RegisterSteps.java:30]: Please change the method name

File: src/test/java/steps/emag/RegisterSteps.java [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+        checkIfTheElementIsNotPresent(RegisterPage.REGISTER_NAME_ERROR_MESSAGE);
+    }
+
+    public static void fillRegisterPasswordField() {
```
Comment (BojteTamas) [inline src/test/java/steps/emag/RegisterSteps.java:34]: Please specify `random` in the name of the method.

File: src/test/java/steps/emag/RegisterSteps.java [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+        waitUntilWebElementIsClickable(RegisterPage.REGISTER_PASSWORD_INPUT);
+        fillWebElement(RegisterPage.REGISTER_PASSWORD_INPUT, password);
+    }
+
+    public static void checkRegisterPasswordFieldInput () {
+        String registerPasswordInputValue = getAttributeFromWebElement(RegisterPage.REGISTER_PASSWORD_INPUT, "value");
+        Assert.assertEquals(password, registerPasswordInputValue);
+    }
+
+    public static void checkRegisterPasswordErrorMessage () {
```
Comment (BojteTamas) [inline src/test/java/steps/emag/RegisterSteps.java:45]: Please change the name  of the method

File: src/test/java/helpers/WebDriverHelpers.java [hunk 17]
```diff
@@ -92,4 +94,24 @@ public static void forceClickWebElement (By byObject) {
         js.executeScript("arguments[0].click()", chromeDriver.findElement(byObject));
```
Comment (BojteTamas) [inline src/test/java/helpers/WebDriverHelpers.java:94]: Please use something like javascriptClick

File: src/test/java/steps/emag/RegisterSteps.java [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+    public static void checkRegisterConfirmPasswordInput () {
+        String registerConfirmPasswordInputValue = getAttributeFromWebElement(RegisterPage.REGISTER_CONFIRM_PASSWORD_INPUT, "value");
+        Assert.assertEquals(password, registerConfirmPasswordInputValue);
+    }
+
+    public static void clickAgreeTermsCheckbox () {
+        forceClickWebElement(RegisterPage.REGISTER_AGREE_TERMS_CHECKBOX);
+    }
+
+    public static void checkAgreeTermsError () {
```
Comment (BojteTamas) [inline src/test/java/steps/emag/RegisterSteps.java:63]: Please be more specific

File: src/test/java/helpers/WebDriverHelpers.java (hunk 1 above)
Comment (MariBuleandra) [reply src/test/java/helpers/WebDriverHelpers.java:99]: Done.

File: src/test/java/helpers/WebDriverHelpers.java (hunk 2 above)
Comment (MariBuleandra) [reply src/test/java/helpers/WebDriverHelpers.java:116]: Done.

File: src/test/java/steps/emag/CommonSteps.java (hunk 3 above)
Comment (MariBuleandra) [reply src/test/java/steps/emag/CommonSteps.java:13]: Done.

File: src/test/java/steps/emag/CommonSteps.java (hunk 4 above)
Comment (MariBuleandra) [reply src/test/java/steps/emag/CommonSteps.java:20]: Done.

File: src/test/java/steps/emag/CommonSteps.java (hunk 5 above)
Comment (MariBuleandra) [reply src/test/java/steps/emag/CommonSteps.java:24]: Done.

File: src/test/java/steps/emag/FavouriteSteps.java (hunk 6 above)
Comment (MariBuleandra) [reply src/test/java/steps/emag/FavouriteSteps.java:18]: Done.

File: src/test/java/steps/emag/LoginSteps.java (hunk 7 above)
Comment (MariBuleandra) [reply src/test/java/steps/emag/LoginSteps.java:30]: Done.

File: src/test/java/steps/emag/MainSteps.java (hunk 8 above)
Comment (MariBuleandra) [reply src/test/java/steps/emag/MainSteps.java:9]: Done.

File: src/test/java/steps/emag/MainSteps.java (hunk 9 above)
Comment (MariBuleandra) [reply src/test/java/steps/emag/MainSteps.java:22]: Done.

File: src/test/java/steps/emag/MainSteps.java (hunk 10 above)
Comment (MariBuleandra) [reply src/test/java/steps/emag/MainSteps.java:26]: Done.

File: src/test/java/steps/emag/MainSteps.java (hunk 11 above)
Comment (MariBuleandra) [reply src/test/java/steps/emag/MainSteps.java:28]: Done.

File: src/test/java/steps/emag/PDPSteps.java (hunk 12 above)
Comment (MariBuleandra) [reply src/test/java/steps/emag/PDPSteps.java:6]: Done.

File: src/test/java/steps/emag/PDPSteps.java (hunk 13 above)
Comment (MariBuleandra) [reply src/test/java/steps/emag/PDPSteps.java:8]: Done.

File: src/test/java/steps/emag/RegisterSteps.java (hunk 14 above)
Comment (MariBuleandra) [reply src/test/java/steps/emag/RegisterSteps.java:30]: Done.

File: src/test/java/steps/emag/RegisterSteps.java (hunk 15 above)
Comment (MariBuleandra) [reply src/test/java/steps/emag/RegisterSteps.java:34]: Done.

File: src/test/java/steps/emag/RegisterSteps.java (hunk 16 above)
Comment (MariBuleandra) [reply src/test/java/steps/emag/RegisterSteps.java:45]: Done.

File: src/test/java/helpers/WebDriverHelpers.java (hunk 17 above)
Comment (MariBuleandra) [reply src/test/java/helpers/WebDriverHelpers.java:94]: Done.

File: src/test/java/steps/emag/RegisterSteps.java (hunk 18 above)
Comment (MariBuleandra) [reply src/test/java/steps/emag/RegisterSteps.java:63]: Done.

File: src/test/java/steps/emag/FavoriteSteps.java [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+
+public class FavoriteSteps extends WebDriverHelpers {
+    static String favoriteProductCount = "3";
+
+    public static void checkIfFavouritesPageUrlIsCorrect () {
+        checkIfTheCurrentUrlContainsText("favorites");
+
+    }
+
+    public static void checkIfFavouritesPageTitleContainsCorrectFavoriteProductCount() {
```
Comment (BojteTamas) [inline src/test/java/steps/emag/FavoriteSteps.java:15]: Here please add the number as parameter, because you cannot change it from the test.

Comment (BojteTamas) [reply src/test/java/steps/emag/FavoriteSteps.java:15]: the favoriteProductCount will remain the same

File: src/test/java/steps/emag/LoginSteps.java [hunk 20]
```diff
@@ -6,27 +6,28 @@
 import pageObjects.emag.LoginPage;
 
 public class LoginSteps extends WebDriverHelpers {
+    public static final String WELCOME_MESSAGE = "Salut!";
```
Comment (BojteTamas) [inline src/test/java/steps/emag/LoginSteps.java:9]: very nice!

File: src/test/java/steps/emag/PdpSteps.java [hunk 21]
```diff
@@ -0,0 +1,45 @@
+package steps.emag;
+
+import helpers.WebDriverHelpers;
+import pageObjects.emag.PdpPage;
+
+public class PdpSteps extends WebDriverHelpers {
+
+    public static final String ADD_TO_CART_MODAL_TITLE = "Produsul a fost adaugat in cos";
```
Comment (BojteTamas) [inline src/test/java/steps/emag/PdpSteps.java:8]: nice!

File: src/test/java/steps/emag/RegisterSteps.java [hunk 22]
```diff
@@ -6,65 +6,66 @@
 import pageObjects.emag.RegisterPage;
 
 public class RegisterSteps extends WebDriverHelpers {
+    public static final String REGISTER_WELCOME_MESSAGE = "Se pare ca nu ai un cont eMAG. Hai sa iti cream un cont nou!";
```
Comment (BojteTamas) [inline src/test/java/steps/emag/RegisterSteps.java:9]: very nice

File: src/test/java/steps/emag/FavoriteSteps.java (hunk 19 above)
Comment (MariBuleandra) [reply src/test/java/steps/emag/FavoriteSteps.java]: Done.

**Pull request merged**